### PR TITLE
double tap to seek without animations

### PR DESCRIPTION
**Pull request by alesmaye**
## Description

We can now seek by double tapping on the UI

## Checklist (for PR submitter and reviewers)

- [x] `CHANGELOG` entry

### CODE REVIEW

File: src/scss/skin-super-modern/components/_smallcenteredplaybacktogglebutton.scss [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    height: 100%;
+    width: 100%;
+  }
+
+  &.#{$prefix}-on {
+    .#{$prefix}-image {
+      background-image: url('../../assets/skin-super-modern/images/pause.svg');
+      //animation: #{$prefix}-fade-out $animation-duration cubic-bezier(.55, .055, .675, .19); // http://easings.net/de#easeInCubic
+      //transition: visibility 0s $animation-duration;
+      //visibility: hidden;
```
Comment (bitmovin-kenny) [inline src/scss/skin-super-modern/components/_smallcenteredplaybacktogglebutton.scss]: Please remove the commented out code

File: src/scss/skin-super-modern/components/_smallcenteredplaybacktogglebutton.scss [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    }
+  }
+
+  &.#{$prefix}-off {
+    .#{$prefix}-image {
+      background-image: url('../../assets/skin-super-modern/images/play.svg');
+      //animation:  #{$prefix}-fade-in $animation-duration cubic-bezier(.55, .055, .675, .19); // http://easings.net/de#easeInCubic
+      //visibility: visible;
```
Comment (bitmovin-kenny) [inline src/scss/skin-super-modern/components/_smallcenteredplaybacktogglebutton.scss]: again, please remove commented out code

File: src/scss/skin-super-modern/components/_uicontainer.scss [hunk 3]
```diff
@@ -34,7 +34,7 @@
     .#{$prefix}-ui-hugeplaybacktogglebutton {
       &:focus {
         box-shadow: inset -4px -3px 2px 9px $color-focus;
-
+
```
Comment (bitmovin-kenny) [inline src/scss/skin-super-modern/components/_uicontainer.scss]: please revert this change since it seems unnecessary

File: src/ts/components/smallcenteredplaybacktogglebutton.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    this.onClick.subscribe(() => {
+      togglePlayback();
+
+      if (firstPlay && this.config.enterFullscreenOnInitialPlayback) {
+        player.setViewMode(player.exports.ViewMode.Fullscreen);
+      }
+    });
+
+    player.on(player.exports.PlayerEvent.Play, () => {
+      // Playback has really started, we can disable the flag to switch to normal toggle button handling
```
Comment (bitmovin-kenny) [inline src/ts/components/smallcenteredplaybacktogglebutton.ts]:
```suggestion
    player.on(player.exports.PlayerEvent.Playing, () => {
      // Playback has really started, we can disable the flag to switch to normal toggle button handling
```
Play only signals the intent to start playback, not actual success.

File: src/ts/components/smallcenteredplaybacktogglebutton.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+      firstPlay = false;
+    });
+
+    player.on(player.exports.PlayerEvent.Warning, (event: WarningEvent) => {
+      if (event.code === player.exports.WarningCode.PLAYBACK_COULD_NOT_BE_STARTED) {
+        // if playback could not be started, reset the first play flag as we need the user interaction to start
+        firstPlay = true;
+      }
+    });
```
Comment (bitmovin-kenny) [inline src/ts/components/smallcenteredplaybacktogglebutton.ts]: This should not matter when we change the code above to wait for playing.

File: src/ts/components/smallcenteredplaybacktogglebutton.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+      this.onToggle.subscribeOnce(() => {
+        this.setTransitionAnimationsEnabled(true);
+      });
+    };
+
+    // Hide the play button animation when the UI is loaded (it should only be animated on state changes)
+    suppressPlayButtonTransitionAnimation();
+
+    const isAutoplayEnabled = player.getConfig().playback && Boolean(player.getConfig().playback.autoplay);
+    // We only know if an autoplay attempt is upcoming if the player is not yet ready. It the player is already ready,
```
Comment (bitmovin-kenny) [inline src/ts/components/smallcenteredplaybacktogglebutton.ts]:
```suggestion
    // We only know if an autoplay attempt is upcoming if the player is not yet ready. If the player is already ready,
```

File: src/ts/components/smallcenteredplaybacktogglebutton.ts [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+      // Enable the transition animations for the next state change
+      this.onToggle.subscribeOnce(() => {
+        this.setTransitionAnimationsEnabled(true);
+      });
+    };
+
+    // Hide the play button animation when the UI is loaded (it should only be animated on state changes)
+    suppressPlayButtonTransitionAnimation();
+
+    const isAutoplayEnabled = player.getConfig().playback && Boolean(player.getConfig().playback.autoplay);
```
Comment (bitmovin-kenny) [inline src/ts/components/smallcenteredplaybacktogglebutton.ts]:
```suggestion
    const isAutoplayEnabled = player.getConfig().playback && Boolean(player.getConfig().playback?.autoplay);
```
There is not necessarily a playback config.

File: src/ts/components/touchcontroloverlay.ts [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+import { UIInstanceManager } from '../uimanager';
+import { EventDispatcher, NoArgs, Event as EDEvent } from '../eventdispatcher';
+import { Timeout } from '../timeout';
+import { HTMLElementWithComponent } from '../dom';
+
+export interface TouchControlOverlayConfig extends ContainerConfig {
+  /**
+   * Specify whether the player should be set to enter fullscreen by clicking on the playback toggle button
+   * when initiating the initial playback.
+   * Default is false.
```
Comment (bitmovin-kenny) [inline src/ts/components/touchcontroloverlay.ts]:
```suggestion
   * Default: false
```
just to be consistent with the other default declarations

File: src/ts/components/touchcontroloverlay.ts [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    });
+
+    this.touchControlEvents.onSingleClick.subscribe((_, e) => {
+      uimanager.getUI().toggleUiShown();
+      playerSeekTime = player.getCurrentTime();
+
+      const eventTarget = (e as Event).target as HTMLElementWithComponent;
+      const rect = eventTarget.getBoundingClientRect();
+      const eventTapX = ((<any>e).clientX) - rect.left;
+      const eventTapY = ((<any>e).clientY) - rect.top;
```
Comment (bitmovin-kenny) [inline src/ts/components/touchcontroloverlay.ts]: Isn't there a better cast than `any`? Maybe `MouseEvent`?

File: src/ts/components/touchcontroloverlay.ts [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+      const eventTarget = event.target as HTMLElementWithComponent;
+      if (!eventTarget || !((eventTarget).component instanceof TouchControlOverlay)) {
+        return;
+      }
+
+      const width = eventTarget.clientWidth;
+      const tapMargin = width * 0.4;
+      const rect = eventTarget.getBoundingClientRect();
+      const eventTapX = ((<any>e).clientX) - rect.left;
+      const eventTapY = ((<any>e).clientY) - rect.top;
```
Comment (bitmovin-kenny) [inline src/ts/components/touchcontroloverlay.ts]: Same here, there should be something better than `any`.
```suggestion
      const eventTapX = ((e as MouseEvent).clientX) - rect.left;
      const eventTapY = ((e as MouseEvent).clientY) - rect.top;
```

File: src/ts/components/touchcontroloverlay.ts [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+        }
+        else if (eventTapX > (width - tapMargin)) {
+          this.touchControlEvents.onSeekForward.dispatch(this);
+        }
+      this.latestTapPosition = {x: eventTapX, y: eventTapY};
+    });
+
+    const thisDomElement = this.getDomElement();
+    thisDomElement.on('click', (e) => {
+      if ((e.target as HTMLElementWithComponent).component instanceof TouchControlOverlay) {
```
Comment (bitmovin-kenny) [inline src/ts/components/touchcontroloverlay.ts:145]: you already did `const eventTarget = event.target as HTMLElementWithComponent` before soe
```suggestion
      if (eventTarget.component instanceof TouchControlOverlay) {
```
could get rid of the cast.

File: src/ts/components/touchcontroloverlay.ts [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+          this.touchControlEvents.onSeekBackward.dispatch(this);
+        }
+        else if (eventTapX > (width - tapMargin)) {
+          this.touchControlEvents.onSeekForward.dispatch(this);
+        }
+      this.latestTapPosition = {x: eventTapX, y: eventTapY};
+    });
+
+    const thisDomElement = this.getDomElement();
+    thisDomElement.on('click', (e) => {
```
Comment (bitmovin-kenny) [inline src/ts/components/touchcontroloverlay.ts]:
```suggestion
    this.getDomElement().on('click', (e) => {
```

File: src/ts/components/touchcontroloverlay.ts [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+      const width = eventTarget.clientWidth;
+      const tapMargin = width * 0.4;
+      const rect = eventTarget.getBoundingClientRect();
+      const eventTapX = ((<any>e).clientX) - rect.left;
+      const eventTapY = ((<any>e).clientY) - rect.top;
+      if (this.latestTapPosition.x === -1 && this.latestTapPosition.y === -1) {
+        this.latestTapPosition = {x: eventTapX, y: eventTapY};
+      }
+      const doubleTapMargin = this.config.seekDoubleTapMargin;
+      if (Math.abs(this.latestTapPosition.x - eventTapX) <= doubleTapMargin && Math.abs(this.latestTapPosition.y - eventTapY) <= doubleTapMargin)
```
Comment (bitmovin-kenny) [inline src/ts/components/touchcontroloverlay.ts]: Does the first click always trigger thsi, since you update the `this.latestTapPosition` to the `eventTapX` value and then `Math.abs(this.latestTapPosition.x - eventTapX) <= doubleTapMargin` should always be true?

**Review by bitmovin-kenny — CHANGES_REQUESTED** [review]
Looks nice overall 👍 some minor issues though.
Only reviewed the code for the first round and didn't test it yet.

File: src/scss/skin-super-modern/components/_uicontainer.scss (hunk 3 above)
Comment (alesmaye) [reply src/scss/skin-super-modern/components/_uicontainer.scss]: this was just a linter warning fix

File: src/ts/components/smallcenteredplaybacktogglebutton.ts (hunk 4 above)
Comment (alesmaye) [reply src/ts/components/smallcenteredplaybacktogglebutton.ts]: This was from the old `HugePlaybackToggleButton`. But yeah... makes sense

File: src/ts/components/smallcenteredplaybacktogglebutton.ts (hunk 7 above)
Comment (alesmaye) [reply src/ts/components/smallcenteredplaybacktogglebutton.ts]: Already checking it with `player.getConfig().playback && ...` but i will change it to `const isAutoplayEnabled = Boolean(player.getConfig().playback?.autoplay);`

File: src/ts/components/touchcontroloverlay.ts (hunk 11 above)
Comment (alesmaye) [reply src/ts/components/touchcontroloverlay.ts:145]: yeah, but not in the scope of the click listener

File: src/ts/components/touchcontroloverlay.ts (hunk 13 above)
Comment (alesmaye) [reply src/ts/components/touchcontroloverlay.ts]: With the changes i made, the first if is not necessary anymore. Already removed

File: src/ts/components/touchcontroloverlay.ts (hunk 11 above)
Comment (bitmovin-kenny) [reply src/ts/components/touchcontroloverlay.ts:145]: Oooohhh right. that is a different one, my bad.

File: src/scss/skin-super-modern/_skin-modern-smallscreen.scss [hunk 14]
```diff
@@ -42,17 +42,19 @@
   }
 
   // Decrease huge play button size and replace icon with normal play icon
-  .#{$prefix}-ui-hugeplaybacktogglebutton {
+  .#{$prefix}-ui-hugeplaybacktogglebutton,
+  .#{$prefix}-ui-smallcenteredplaybacktogglebutton {
     .#{$prefix}-image {
-      background-image: url('../../assets/skin-super-modern/images/play.svg');
+      //background-image: url('../../assets/skin-super-modern/images/play.svg');
```
Comment (bitmovin-kenny) [inline src/scss/skin-super-modern/_skin-modern-smallscreen.scss]: one more commented line

**Review by bitmovin-kenny — APPROVED** [review]
Nice
